pkp/pkp-lib#983 improve reviewer selection

NateWr opened the pull request:
See #983 for further comments.

asmecher commented on styles/controllers/rangeSlider.less:
```diff
@@ -11,24 +11,85 @@
+		width: @base;
+		height: @base;
+		margin-left: -@half;
+		background-image: url('data:image/svg+xml;base64,PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0idXRmLTgi…gd2lkdGg9IjEwMCUiIGhlaWdodD0iMTAwJSIgZmlsbD0idXJsKCNncmFkKSIgLz48L3N2Zz4g');
```
I haven't worked with base64-encoded images in CSS before. How maintainable is this? Is it now common practice?

NateWr replied:
To be honest, we can probably get rid of it. I pulled this style out from a codepen and the `background-image` is a fallback for browsers that don't support linear gradients (IE 9). It's probably not clear to subsequent maintainers that it would need to change if the style changed. So probably not a great idea. And IE 9 users will still see a flat white handle, which won't be terrible.

Note to self: there's another base-64 encoded image below in the :after element.

asmecher commented on templates/form/link.tpl:
```diff
@@ -7,7 +7,4 @@
-<div{if $FBV_layoutInfo} class="{$FBV_layoutInfo}"{/if}>
-	<a href="{$FBV_href}" id="{$FBV_id}" class="{$FBV_class}">{translate key=$FBV_label}</a>
-</div>
+<a href="{$FBV_href}" id="{$FBV_id}" class="{$FBV_class}">{translate key=$FBV_label}</a>
```
(This template doesn't appear to broadly used; I suggest we put it on the chopping block.)

NateWr commented:
I've committed a couple things addressing the comments regarding base64 encoded images and getting rid of `link.tpl`. I'll wait to update the OMP PR until we resolve the issues raised by the tests there. https://github.com/pkp/omp/pull/196

asmecher commented on classes/user/PKPUserDAO.inc.php:
```diff
@@ -230,6 +230,7 @@ function getAllReviewers($contextId) {
 	/**
 	 * Given the ranges selected by the editor, produce a filtered list of reviewers
 	 * @param $contextId int
+	 * @param $name string|null Partial string match with reviewer name
```
Looks like there are some missing parameter docs here, but not from your addition, Nate. I'll add this when I merge.

asmecher commented on classes/user/PKPUserDAO.inc.php:
```diff
@@ -243,7 +244,7 @@ function getAllReviewers($contextId) {
 	 * @return array Users
 	 */
-	function getFilteredReviewers($contextId, $stageId, $doneMin = null, $doneMax = null, $avgMin = null, $avgMax = null, $lastMin = null, $lastMax = null, $activeMin = null, $activeMax = null, $interests = array(), $submissionId = null, $reviewRoundId = null) {
+	function getFilteredReviewers($contextId, $stageId, $name = null, $doneEnabled = null, $doneMin = null, $doneMax = null, $avgEnabled = null, $avgMin = null, $avgMax = null, $lastEnabled = null, $lastMin = null, $lastMax = null, $activeEnabled = null, $activeMin = null, $activeMax = null, $interests = array(), $submissionId = null, $reviewRoundId = null) {
```
This parameter list is getting pretty unmanageable; why not check `$lastMin !== null || $lastMax !== null` to imply whether `$lastEnabled` (similar with the others)?

NateWr replied:
I agree it's an obnoxious line of code.

As I understand it, the `lastMin` and `lastMax` values are passed along in the form submission, even if `lastEnabled` is checked. I can see two approaches:

1. We strip out the values in JavaScript before they're posted to the server, if the field is disabled. This will require a lot of specific code in `ClientFormHandler.js` which we probably don't want.

2. Right around [here](https://github.com/pkp/pkp-lib/pull/1064/files#diff-66ad5c92e421039f092ebca5dc27a26aR163) we reset the min/max values to `null` if the enabled value is false.

The second seems better, yea?

asmecher replied:
Yes, the second approach would make it somewhat better.

I've mused about the possibility of creating and passing in some kind of object to describe filtering, but let's not do that just yet.

asmecher commented on styles/form/form.less:
```diff
@@ -394,28 +394,44 @@
 }
 
+.ui-autocomplete-loading {
+	background-image: url(data:image/gif;base64,R0lGODlhHgAeAPf2AP7+/v39/fDw8O/v7/z8/PHx8e7u7vv7++Xl5fr6+vn5+ebm5gAAAPX19fT09Pb29vPz8/f39/j4+Ofn5/Ly8tTU1O3t7dXV1cnJyezs7Ojo6Orq6uTk5OPj476+vuvr69nZ2cjIyNbW1unp6crKytjY2MvLy9zc3LOzs7KyssfHx+Hh4b+/v9/f3+Li4tPT097e3sDAwNfX193d3dra2sHBwYmJidvb2+Dg4L29vby8vM/Pz7e3t9LS0sTExNDQ0LS0tIiIiLW1tcbGxszMzLi4uLq6uoyMjHBwcMPDw8XFxVhYWLGxsXFxccLCws7Ozra2trCwsG9vb42Njb ...
```
I think you were planning to remove base64-encoded images in CSS?

NateWr replied:
Ah yeah, this went in before we had this discussion. This one is more valuable (not just an IE9 fallback) and wouldn't need to be updated alongside other CSS changes (to e.g. the background color).

I can remove it if you want, but I'll need to go hunting for an open source animated .gif to use. Usually we would use :psuedo elements with CSS animations for this kind of thing, but we can't attach them to `input` fields so we need a background image.

asmecher replied:
I only pointed it out because it's new to me. I don't mind using it.

NateWr replied:
Ok let's leave this one until someone complains. :)

asmecher commented on templates/form/formButtons.tpl:
```diff
@@ -30,9 +30,9 @@
 		{/if}
 		{if $FBV_formReset}
-			{fbvElement type="link" class="resetButton" id="resetButton"|concat:"-":uniqid label=$FBV_cancelText}
+			<a href="#" id="resetButton-{$uniqid}" class="resetButton">{translate key=$FBV_cancelText}</a>
```
Is the change from {uniqid} (the function) to {$uniqid} (the variable) intentional?

NateWr replied:
No, I didn't realise it was calling a function there. Do you know the `smarty` syntax for calling that function in that place?

asmecher replied:
Hmm, it's verbose, but...

    {assign var=resetButtonId value="resetButton-"|uniqid}
    <a href="#" id="{$resetButtonId}" ...

...should do it.

NateWr replied:
Ok I've chucked that in there but can't actually find a use of the reset button to test with. Maybe it can be removed, along with the related code in `FormHandler.js`?

asmecher replied:
Agreed, it wasn't used anywhere so I cleaned out that code/markup.

NateWr commented:
Looks like there are merge conflicts. I'm finishing for the day in a few minutes. And I'm travelling to London tomorrow for a conference. I'll try to get PRs for this updated and ready to go tomorrow.

asmecher commented:
Replaced by https://github.com/pkp/pkp-lib/pull/1095
